### PR TITLE
fix flicker in the popup menu by SolaWing patch

### PR DESCRIPTION
**Pull request by Shougo**
It is the neovim patch convertion from https://github.com/neovim/neovim/pull/2674 by @SolaWing.

It fixes the popup menu flick using `'refresh':'always'` feature like YouCompleteMe plugin.

Reproduce ways:

```
* $ vim -Nu test.vim
* Enter the insert mode
* Input "Ja" characters
* You can see the flick if the patch is not available
```

test.vim

``` vim
fun! CompleteMonths(findstart, base)
  if a:findstart
    " locate the start of the word
    let line = getline('.')
    let start = col('.') - 1
    while start > 0 && line[start - 1] =~ '\a'
      let start -= 1
    endwhile
    return start
  else
    " find months matching with "a:base"
    let ret = []
    for m in split("January February March April May June July Auguest September October November December")
      if m =~ '^' . a:base
        call add(ret, m)
      endif
    endfor
    echom "return words"
    sleep 400m  " simulate searching for next match
    return {'words':ret, 'refresh':'always'}
  endif
endfun

set completeopt=menuone,noselect
set completefunc=CompleteMonths
set omnifunc=CompleteMonths
aug mytest
  au!
  au CursorMovedI <buffer> echom 'moveed' | call feedkeys("\<C-x>\<C-u>", 'n')
aug end
```

### CODE REVIEW

**Comment by justinmk** [comment]
Rebase will fix the failing test.

**Comment by Shougo** [comment]
Thank you for the comment!  Rebased.

**Comment by brammool** [comment]
Shougo wrote:

> It is the neovim patch convertion from https://github.com/neovim/neovim/pull/2674 by @SolaWing.
> 
> It fixes the popup menu flick using `'refresh':'always'` feature like
> YouCompleteMe plugin.

With "flick" you mean flicker?

> Reproduce ways:
> 
> ```
> * $ vim -Nu test.vim
> * Enter the insert mode
> * Input "Ja" characters
> * You can see the flick if the patch is not available
> ```

The logic is very hard to follow.  Mostly caused by existing code.  For
example, the use of variables that don't have any comment to explain how
they are used.

After the change the compl_opt_refresh_always variable is used with:

```
if ((ctrl_x_mode == CTRL_X_FUNCTION || ctrl_x_mode == CTRL_X_OMNI)
    && compl_opt_refresh_always)
```

What other value of ctrl_x_mode is there when compl_opt_refresh_always
can be TRUE?

I can't say by looking at the code whether this patch will cause trouble
in some situations.  Did several people check it out?

## 

hundred-and-one symptoms of being an internet addict:
232. You start conversations with, "Have you gotten an ISDN line?"

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\
\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by SolaWing** [comment]
@brammool 

> With "flick" you mean flicker?

Yes, it was just a typo.

> if ((ctrl_x_mode == CTRL_X_FUNCTION || ctrl_x_mode == CTRL_X_OMNI)
>       && compl_opt_refresh_always)
> 
> What other value of ctrl_x_mode is there when compl_opt_refresh_always
> can be TRUE?

Yes, the `compl_opt_refresh_always` missed setting to False sometimes. Reproduce by:
- Complete With `CTRL_X_FUNCTION` or `CTRL_X_OMNI`, and set `compl_opt_refresh_always` to true
- before previous completion end, input `<C-X><C-N>` to change completion mode.
- now `compl_opt_refresh_always` is always true.

For safety, I add ctrl_x_mode check.

The detail code change reason I explained in neovim/neovim#2674

**Comment by Shougo** [comment]
@brammool @SolaWing 

Thank you!

I have rebased and fixed some minor changes.
- `flick` -> `flicker`
- use `ins_comp_need_restart()` instead of `ctrl_x_mode` and `compl_opt_refresh_always` checks

**Comment by demon386** [comment]
This solves my long time annoyance with using Vim + YouCompleteme in tmux.

**Comment by Shougo** [comment]
I have updated the fix.

Included the change.
https://github.com/SolaWing/neovim/commit/3803198f80c27b226ca8ff7ffbc015ce341f5385

**Comment by amadeus** [comment]
Looks like this PR hasn't been touched in a while, any way we can get it merged in?

**Comment by alexander-akhmetov** [comment]
> Looks like this PR hasn't been touched in a while, any way we can get it merged in?

+1 :)

**Comment by brammool** [comment]
This PR no longer looks valid.  Removing the redo buffer calls looks wrong anyway.
I'm closing this now, if the fix is still needed please create a new PR.
